main.go

```diff
@@ -23,31 +28,90 @@ import (
 var version = "dev"
 var date = "2006-01-02T15:04:05Z"
 
+type debugInfoQuery struct {
+	DebugInfo struct {
+		SelfHostedVersion string `graphql:"selfHostedVersion"`
+	} `graphql:"debugInfo"`
+}
+
+func getSpaceliftInstanceVersion() cmd.SpaceliftInstanceVersion {
+	// Default fallback when we can't determine the version
+	instanceVersion := cmd.SpaceliftInstanceVersion{
+		InstanceType: cmd.SpaceliftInstanceTypeUnknown,
+	}
+
+	ctx, httpClient := session.Defaults()
+
+	// Create a new session - this may fail if the user doesn't have valid credentials
+	sess, err := session.New(ctx, httpClient)
+	if err != nil {
+		return instanceVersion
+	}
+
+	// Create a new client using the session
+	c := client.New(httpClient, sess)
+
+	// Make the GraphQL query to get debug info
+	var query debugInfoQuery
+	if err := c.Query(context.Background(), &query, nil); err != nil {
+		// If the query fails, return the unknown instance type
+		return instanceVersion
+	}
+
+	// If the query succeeds, determine if this is a SaaS or Self-Hosted instance
+	if query.DebugInfo.SelfHostedVersion == "" {
+		// Empty version means SaaS
+		instanceVersion.InstanceType = cmd.SpaceliftInstanceTypeSaaS
+	} else {
+		// Non-empty version means Self-Hosted
+		instanceVersion.InstanceType = cmd.SpaceliftInstanceTypeSelfHosted
+
+		// Parse the version string into a semantic version
+		v, err := semver.NewVersion(query.DebugInfo.SelfHostedVersion)
+		if err == nil {
+			instanceVersion.Version = v
+		} else {
+			log.Printf("Warning: Failed to parse Self-Hosted version string: %q: %v",
+				query.DebugInfo.SelfHostedVersion, err)
+		}
+	}
+
+	return instanceVersion
+}
+
 func main() {
 	compileTime, err := time.Parse(time.RFC3339, date)
 	if err != nil {
 		log.Fatalf("Could not parse compilation date: %v", err)
 	}
+
+	instanceVersion := getSpaceliftInstanceVersion()
+
+	if instanceVersion.InstanceType == cmd.SpaceliftInstanceTypeUnknown {
+		log.Println("Warning: Unable to determine Spacelift instance type. Some commands may be unavailable until you authenticate with Spacelift.")
+	}
+
 	app := &cli.App{
 		Name:                 "spacectl",
 		Version:              version,
 		Compiled:             compileTime,
 		Usage:                "Programmatic access to Spacelift GraphQL API.",
 		EnableBashCompletion: true,
-		Commands: []*cli.Command{
-			module.Command(),
+		Commands: append([]*cli.Command{
```

Comment:
Really good that you thought about these from the start.

Reply:
Yeah - while I was working on this I was worried we'd hit a chicken and egg situation. But I think it ended up working out fine, and we just have some core commands that are always available.